ci-operator/step-registry/cucushift/installer/check/reboot-nodes/cucushift-installer-check-reboot-nodes-commands.sh

```diff
@@ -0,0 +1,114 @@
+set -o pipefail
+
+function count_reboot_event() {
+    local node_list reboot_count
+    node_list=$1
+    reboot_count=0
+
+    for node_name in ${node_list}; do
+        events_number=$(oc get events -n default | awk "/Rebooted/ && /${node_name}/{++c}END{print (c == \"\" ? 0 : c)}")
+        reboot_count=$((reboot_count + events_number))
+    done
+    echo $reboot_count
+}
+
+function ssh_command() {
+    local node_ip="$1"
+    local cmd="$2"
+
+    ssh_options="-o UserKnownHostsFile=/dev/null -o IdentityFile=${SSH_PRIV_KEY_PATH} -o StrictHostKeyChecking=no"
+    ssh_proxy_command=""
+    if [[ -f "${SHARED_DIR}/bastion_public_address" ]]; then
+        bastion_ip=$(<"${SHARED_DIR}/bastion_public_address")
+        bastion_ssh_user=$(<"${SHARED_DIR}/bastion_ssh_user")
+        ssh_proxy_command="-o ProxyCommand='ssh ${ssh_options} -W %h:%p ${bastion_ssh_user}@${bastion_ip}'"
+    fi
+
+    echo "ssh ${ssh_options} ${ssh_proxy_command} core@${node_ip} '${cmd}'" | sh -
+}
+
+function reboot_cluster() {
+    local try max_try reboot_count_before=0 reboot_count_after=0 total_nodes_count=0 master_list node_list="" worker_list
+    declare -A node_ip_array
+    master_list=$(oc get node -o wide --no-headers | grep 'master' | awk '{print $1":"$6}' | sort)
+    if [ X"${SIZE_VARIANT}" == X"compact" ]; then
+        node_list="${master_list}"
+    else
+        worker_list=$(oc get node -o wide --no-headers | grep 'worker' | awk '{print $1":"$6}' | sort)
+        node_info_list="${worker_list} ${master_list}"
+    fi
+    for node_info in ${node_info_list}; do
+        node_name=${node_info/:*}
+        node_ip=${node_info#*:}
+        node_list="${node_list} ${node_name}"
+        node_ip_array[${node_name}]=${node_ip}
+    done
+    total_nodes_count=$(echo ${node_list} | awk '{print NF}')
+    reboot_count_before=$(count_reboot_event "${node_list}")
+    for node_name in ${node_list}; do
+        node_ip=${node_ip_array[${node_name}]}
+        echo "$(date -u --rfc-3339=seconds) - rebooting node ${node_name}, node ip is ${node_ip}"
+        ssh_command "${node_ip}" "sudo reboot"
+        # add some sleep between node reboot to avoid etcd cluster can not record reboot event
+        sleep 120
+    done
+    try=0
+    max_try=6
+    while [[ ${try} -lt ${max_try} ]]; do
+	if [[ $(oc get node --no-headers | grep -c 'Ready') -lt 0 ]]; then
+	    echo "$(date -u --rfc-3339=seconds) - wait for node boot up"
+	else
+	    reboot_count_after=$(count_reboot_event "${node_list}")
+	    temp_count=$(( reboot_count_after - reboot_count_before))
+            echo "DEBUG: total_nodes_count: ${total_nodes_count}, reboot_count_before: ${reboot_count_before}, reboot_count_after: ${reboot_count_after}"
+	    # because etcd can not record reboot event when the whole cluster is rebooting,
+            # so inaccuracy should be controled less than 2
+            if [[ ${total_nodes_count} -gt 0 ]] && [[ $((total_nodes_count - temp_count)) -lt 2 ]]; then
```

Comment:
In v4-installer-autotest, because we were using `oc debug node` to reboot nodes, so we count events to ensure the nodes really rebooted, now we are run ssh command to reboot it.

To be more stable, maybe we do not need to count reboot event, we can find out some alternatives.
e.g:
1. run `sudo reboot -f`, and save its output
2. judge the command return 0
3. judge some keyword in the output to double confrm.
4. after the whole cluster are rebooted, print out events to help later debugging.

Reply:
thanks for the suggestion, updated, please help to review.